Add Subscribers Stats section to the homepage [MAILPOET-4828]

costasovo opened the pull request:
## Description

This PR adds the Subscribers Stats section to the plugin's homepage.

**New install**
<img width="722" alt="Screenshot 2023-01-25 at 15 11 13" src="https://user-images.githubusercontent.com/1082140/214585988-685e49f0-c8d3-48a0-bd93-86be60334a88.png">

**After a few subscribers subscribe**
<img width="714" alt="Screenshot 2023-01-25 at 15 12 31" src="https://user-images.githubusercontent.com/1082140/214586164-16734a17-c896-46de-9a92-0c03a0cf16ac.png">


## Code review notes

I recommend checking the PR commit by commit.

## QA notes

The homepage is still hidden behind a feature flag, so you need to enable it from `wp-admin/admin.php?page=mailpoet-experimental`

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4828]

## After-merge notes

_N/A_


[MAILPOET-4828]: https://mailpoet.atlassian.net/browse/MAILPOET-4828?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Add empty subscribers stats section to homepage

[MAILPOET-4828]

costasovo committed:
Add subscribers section description

[MAILPOET-4828]

costasovo committed:
Add fetching of global subscribers stats to Homepage data controller

[MAILPOET-4828]

costasovo committed:
Display global subscriber stats in the homepage UI

[MAILPOET-4828]

costasovo committed:
Add fetching of list level subscriber changes to homepage data service

[MAILPOET-4828]

costasovo committed:
Render list change info in subscribers stats section on homepage

[MAILPOET-4828]

costasovo committed:
Add engagement score to the homepage subscribers stats

[MAILPOET-4828]

costasovo committed:
Add empty states for homepage subscribers stats section

[MAILPOET-4828]

costasovo committed:
Compute change of global subscribed subscribers

[MAILPOET-4828]

costasovo committed:
Add global subscribers stats change info to UI

[MAILPOET-4828]

costasovo committed:
Add styles for subscribers stats section

[MAILPOET-4828]

costasovo committed:
Add styles for small screens for subscriber stats

[MAILPOET-4828]

costasovo committed:
Add acceptance test for subscribers stats section on the homepage

[MAILPOET-4828]

costasovo committed:
Add description of what subscriber stats values represent

[MAILPOET-4828]

costasovo committed:
Add missing key property for the list of lists

[MAILPOET-4828]

triple0t commented on mailpoet/assets/js/src/homepage/components/subscribers-stats.tsx:
```diff
@@ -0,0 +1,153 @@
+      heading={MailPoet.I18n.t('subscribersHeading')}
+      description={MailPoet.I18n.t('subscribersSectionDescription')}
+      headingAfter={
+        !missingStats ? (
+          <span
+            className={classnames({
+              'mailpoet-decrease': globalChange.changePercent < 0,
+            })}
+          >
+            <Icon icon={trendingUp} />
+            {globalChange.changePercent < 0 ? '-' : ''}
+            {Math.abs(globalChange.changePercent) > 1000
+              ? '∞%'
+              : `${Math.abs(globalChange.changePercent)}%`}
+          </span>
+        ) : null
```
I noticed a number of logical checks in this template.

Rather than using something like `check ? doA : null`
why not use something like: `check && doA`

This would compress the codes a bit and make it easier to read

costasovo replied:
Good point! I personally preffer expressions over ternary operator too. For some reason I thought that our eslint doesn't allow such expressions in react so I automatically went with ternary. Now I see I can use expressions. Hmm I don't know where I got the impression. I'll add a commit with a refactor addressing this.

triple0t commented on mailpoet/assets/js/src/homepage/components/subscribers-stats.tsx:
```diff
@@ -0,0 +1,153 @@
+                ))}
+              </tbody>
+            </table>
+          ) : null}
```
Something similar for this check

triple0t commented on mailpoet/assets/js/src/homepage/components/homepage-sections.tsx:
```diff
@@ -26,6 +27,9 @@ export function HomepageSections(): JSX.Element {
           <ProductDiscovery onHide={hideProductDiscovery} />
         </ErrorBoundary>
       ) : null}
+      <ErrorBoundary>
+        <SubscribersStats />
+      </ErrorBoundary>
```
I have a little understanding of why you are doing this, but I want to confirm. Do we need so many Error Boundaries?

We can also compress this line down to
```suggestion
          {!isProductDiscoveryHidden && (
        <ErrorBoundary>
          <ProductDiscovery onHide={hideProductDiscovery} />
        </ErrorBoundary>
      )}
      <ErrorBoundary>
        <SubscribersStats />
      </ErrorBoundary>
```

costasovo replied:
> I have a little understanding of why you are doing this, but I want to confirm. Do we need so many Error Boundaries?

@triple0t I forgot to answer this question. I thought it might be a good idea to wrap each section into a boundary so that when one section breaks down with an error, the other will still be available for the user. I see those sections as independent widgets so it makes sense to me to wrap them like that. But it is not any super strong opinion.

triple0t replied:
I understand.

triple0t reviewed (commented):
Looks good.

Well done, @costasovo 👏🏾

I noticed an issue with the list table.
<img width="767" alt="Screenshot 2023-01-27 at 8 43 10 AM copy" src="https://user-images.githubusercontent.com/30554163/215037861-4abd7e17-01b9-4694-9745-c6ef77b20ff8.png">

It seems that only subscribed stats are added to the list.
Unsubscribed stats were not.

costasovo committed:
Refactor homepage code - prefer expression over ternary operator

[MAILPOET-4828]

costasovo committed:
Improve global unsubscribes stats on homepage

Before this fix if a subscriber subscribed multiple times
we counted them a multiple global unsubscribes
[MAILPOET-4828]

costasovo committed:
Improve list level unsubscribe stats

We want to show admin how many subscribers subscribed to a list and how many
unsubscribed within last 30 days so that they see the change.
If someone subscribed and also unsubscribed within those 30 days we skip him.
We don't have data to detect they were really subscribed at some point.
[MAILPOET-4828]

costasovo commented:
@triple0t Thank you for the review. I addressed the coding styles issues and added a few fixes/improvements to how we fetch those stats. But I'm not sure my fixes cover the issue from the screenshot.

1) Theoretically this might be a valid state in case the subcriber was globally subscribed but not in any list.
2) Another explanation could be that these subscribers were unsubscribed globally by stayed subscribed on the list level. On unsubscribe we [remove then from lists](https://github.com/mailpoet/mailpoet/blob/b0ae21ec6067f2653778581a85f9d451e702fbfc/mailpoet/lib/Subscription/Pages.php#L277) but we keep WP Users list status. So maybe these 3 unsubscribed were WP users?

Could you please check in you DB if these subribers have some lists in DB and what are the statuses in lists? Maybe Woo or WP user sync that causes this state 🤔.

I want to add that because of the 2 levels of subcription statuses I struggled with fetching the correct numbers and it still feels that it might be confusing for our users. Please see related slack discussion ([#](https://a8c.slack.com/archives/C01GH764202/p1674136218312439)). This may help to uderstand why thigs are way they are. We can still re-open it and chnage the stats. I'm open to any ideas for improvements.

triple0t commented:
> @triple0t Thank you for the review. I addressed the coding styles issues and added a few fixes/improvements to how we fetch those stats. But I'm not sure my fixes cover the issue from the screenshot.
> 
> 1. Theoretically this might be a valid state in case the subcriber was globally subscribed but not in any list.
> 2. Another explanation could be that these subscribers were unsubscribed globally by stayed subscribed on the list level. On unsubscribe we [remove then from lists](https://github.com/mailpoet/mailpoet/blob/b0ae21ec6067f2653778581a85f9d451e702fbfc/mailpoet/lib/Subscription/Pages.php#L277) but we keep WP Users list status. So maybe these 3 unsubscribed were WP users?
> 
> Could you please check in you DB if these subribers have some lists in DB and what are the statuses in lists? Maybe Woo or WP user sync that causes this state 🤔.
> 
> I want to add that because of the 2 levels of subcription statuses I struggled with fetching the correct numbers and it still feels that it might be confusing for our users. Please see related slack discussion ([#](https://a8c.slack.com/archives/C01GH764202/p1674136218312439)). This may help to uderstand why thigs are way they are. We can still re-open it and chnage the stats. I'm open to any ideas for improvements.

I see.

I can confirm the user was unsubscribed on the global level but not the list level. This, indeed explains the discrepancy.

Thanks for the link to the Slack conversation. I will take a look.

I'm approving this PR. Well done.

triple0t reviewed (dismissed):
LGTM 💪🏾

Aschepikov commented:
Hi, @costasovo 
 Links `Explore campaigns` and `Create a subscription form` are redirecting to the homepage again if I want to open them in the new tab.

The link `Explore campaigns` redirects me to the newsletters https://qawp.net/wp-admin/admin.php?page=mailpoet-newsletters#/new but it looks like a link to the existing forms to explore them because it is like the advice "oh you don't have new subscribers in the last 30 days so go and explore your current campaigns and start a new one, but new campaign it is a form and not a newsletter because newsletter I can send just to existed subscribers. We may need to make this link redirecting to the existing forms. What do you think @pavel-mailpoet ?

I don't see unsubscribed count in the changes. 
![image](https://user-images.githubusercontent.com/113185437/216314249-02f8e183-c209-4cd7-9b90-229a45d52608.png)
For example, here on the new install I have +1 subscriber 
![image](https://user-images.githubusercontent.com/113185437/216314538-9bc59a55-2162-41a7-a52b-81da31ade86e.png)
Then I am adding one more subscriber and have +2
![image](https://user-images.githubusercontent.com/113185437/216314650-93804b32-5846-49ef-a292-13383e04e684.png)
Then I unsubscribed the last subscriber and I don't see him in the changes or lists because he was unsubscribed from the list and doesn't have the list in the table 
![image](https://user-images.githubusercontent.com/113185437/216315896-a2299354-1071-4358-af13-ceb936b48812.png)
I tried with unsubscribing as admin and as a client from the newsletter unsubscribe link
![image](https://user-images.githubusercontent.com/113185437/216315139-f9892e06-48d9-4cd9-8c74-8f94136baeee.png)

Also If I will make him unsubscribed but with a list in the table (I am not sure that I can do this like a client) 
![image](https://user-images.githubusercontent.com/113185437/216316275-03a3c62c-5db7-4658-8e02-154d817e30e5.png)
I wouldn't see him in the unsubscribed on the homepage
![image](https://user-images.githubusercontent.com/113185437/216316348-c889f496-83b5-4999-b38b-d1b308b99823.png)
But when I am redirecting with a list link from the homepage I can see him 
![image](https://user-images.githubusercontent.com/113185437/216316588-173723a4-00c8-4927-aaf3-5afc249753c4.png)

And there is a question why I can see this unsubscribed here in the lists,I cant reproduce them and the count is incorrect. 
![image](https://user-images.githubusercontent.com/113185437/216318951-8e9eab5f-bcd3-4b28-a52d-139f2280d257.png)
Here in the newsletter mailing list I have 3 subscribed and 1 unsubscribed in the last 30 days, but if I will go to the list link I will see 2 unsubscribed 
![image](https://user-images.githubusercontent.com/113185437/216319269-8b544f8f-0f83-44a0-b6ba-5681aea8592d.png)
Maybe it is connected with the date and today is not showing on the unsubscribe stats but is showing in the subscribe. But the total amount is still wrong with this. And I tried to unsubscribe as a user too from our unsubscribe links in the email. Can you check it pls, if you will need I can prepare these cases with screenshots from db with all our methods of unsubscribing.

pavel-mailpoet commented:
> The link Explore campaigns redirects me to the newsletters https://qawp.net/wp-admin/admin.php?page=mailpoet-newsletters#/new but it looks like a link to the existing forms to explore them because it is like the advice "oh you don't have new subscribers in the last 30 days so go and explore your current campaigns and start a new one, but new campaign it is a form and not a newsletter because newsletter I can send just to existed subscribers. We may need to make this link redirecting to the existing forms. What do you think @pavel-mailpoet ?

It is correct. The link should go to the newsletters page. User can see the link only if they have already an active form and they didn't get any subscribers for the last 30 days. I reckon this is a rare situation. If they already have a form it is not very practical to send the user to the forms page. So sending them to the newsletters seem like the next best option.

costasovo committed:
Add links for empty states of subscribers stats to allow open in new tab

[MAILPOET-4828]

Aschepikov commented:
> > The link Explore campaigns redirects me to the newsletters https://qawp.net/wp-admin/admin.php?page=mailpoet-newsletters#/new but it looks like a link to the existing forms to explore them because it is like the advice "oh you don't have new subscribers in the last 30 days so go and explore your current campaigns and start a new one, but new campaign it is a form and not a newsletter because newsletter I can send just to existed subscribers. We may need to make this link redirecting to the existing forms. What do you think @pavel-mailpoet ?
> 
> It is correct. The link should go to the newsletters page. User can see the link only if they have already an active form and they didn't get any subscribers for the last 30 days. I reckon this is a rare situation. If they already have a form it is not very practical to send the user to the forms page. So sending them to the newsletters seem like the next best option.

But @pavel-mailpoet  how it can help me? As I understand Campaign here is something that can help me to grow my audience or already helped/not helped. I can do this in our plugin just with forms. For example, I have a shop with some little handmade things, and I have a form that is really badly designed and coupon with some shit stuff, so nobody wants to be my subscriber and I don't have subscribers last 30 days, and it will be logically that our stats in the homepage will tell me "Your subscriber count hasn’t changes in the last 30 days. Launch a new campaign to grow your audience." and a button to create this campaign/form or to explore existed campaigns to understand why they are not working
![image](https://user-images.githubusercontent.com/113185437/216325828-47204fb9-cd03-47d9-9eda-cf814e28d208.png)
When I am looking at our types of newsletters I don't understand how they are connected with the growing of my audience. 
I am not arguing so if I am wrong ok, its a part of my job to be sure in the quality of customer experience and I need to ask these things, you know mailpoet better and if you think that newsletters are better so let it be, it means that I need to know our product more and that's it.

costasovo commented:
> But @pavel-mailpoet how it can help me? As I understand Campaign here is something that can help me to grow my audience or already helped/not helped. I can do this in our plugin just with forms. For example, I have a shop with some little handmade things, and I have a form that is really badly designed and coupon with some shit stuff, so nobody wants to be my subscriber and I don't have subscribers last 30 days, and it will be logically that our stats in the homepage will tell me "Your subscriber count hasn’t changes in the last 30 days. Launch a new campaign to grow your audience." and a button to create this campaign/form or to explore existed campaigns to understand why they are not working

@Aschepikov I had the very same thoughts when working on the ticket.

@pavel-mailpoet I think that the meaning "explore campaigns" fits much better to "explore existing campaigns (aka forms) to understand why they are not working". Maybe we could alter the wording somehow?

pavel-mailpoet commented:
I think you are right. It makes sense to send users to the forms page. So renaming the button to "Explore your forms" and sending a user to the existing forms page would be the best. 

Campaigns are newsletters so we shouldn't mix those two.

costasovo commented:
@Aschepikov Thanks for the feedback!

> Links Explore campaigns and Create a subscription form are redirecting to the homepage again if I want to open them in the new tab.

Good catch! I fixed it in a new commit.

The subscribers count is a bit more complex problem. It is kind of related to the two levels of subscription status (global and list level) that we have in the plugin. I'll try to explain why things are the way they are. I'm open to all ideas on how we could improve this because I think the numbers might be confusing for our users. Please see this comment with an explanation how [we count the numbers](https://github.com/mailpoet/mailpoet/pull/4683/commits/1a99e0422a7466553e84626d5e69176851780de8) and also [here is a slack discussion about the problem](https://a8c.slack.com/archives/C01GH764202/p1674136218312439).

I'll try to explain the issues you reported.

> 1) I don't see unsubscribed count in the changes.

This is related to list status VS global status. A subscriber might be removed from the list but still remain globally subscribed.  For example, when a subscriber who is subscribed to three lists visits the manage subscription page and remove themselves from two lists, there will be two records in the lists stats table, but zero in the global table.

> 2) -For example, here on the new install I have +1 subscriber
-Then I am adding one more subscriber and have +2
-Then I unsubscribed the last subscriber and I don't see him in the changes or lists because he was unsubscribed from the list and doesn't have the list in the table
-I tried with unsubscribing as an admin and as a client from the newsletter unsubscribe link
-Here in the newsletter mailing list I have 3 subscribed and 1 unsubscribed in the last 30 days, but if I will go to the list link I will see 2 unsubscribed

I think that all these are related to the fact that in numbers, I exclude subscribers who subscribed and also unsubscribed within the last 30 days. That's why you had (+2/-0), and after you unsubscribed the one, you went to (+1/-0) but I assume you would expect (+2/-1). The one who subscribed and unsubscribed is ignored. 

I thought that it might be a good idea to count it that way because, in these stats, we display the overall change. Also, it is a bit tricky to count those who both subscribed and unsubscribed in the last 30 days into the subscribed stats because their current status is unsubscribed.
But I see that this behavior might be confusing especially when testing the functionality. So the other way around might be a better option here. We have `last_subscribed_at` in the subscribers table, so I think we can use this column instead of the current status to detect who subscribed in the last 30 days.
On the list level, we currently create the record only when a user subscribes, so I guess I can use `created_at` and again ignore the current status to get the subscribed in the last 30 days' stats that includes also those who unsubscribed.

With the current implementation when 10 subscribers subscribe and 9 of them unsubscribe within 30 days we display (+1/-0) with the proposed change we would display (+10/-9)

@pavel-mailpoet I would like to confirm with you that the change I propose ⬆️ makes sense to you. Does it? Do you see some potential issues?

pavel-mailpoet commented:
> I would like to confirm with you that the change I propose ⬆️ makes sense to you. Does it? Do you see some potential issues?

I think it makes sense. I am not against changing it. I reckon when we send it to real users, we will see how confusing/clear it really is.

costasovo committed:
Add helper method for updating updatedAt of an entity in integration tests

We can't simply set updatedAt via setter on the entity. The value gets overwritten
in the TimeStampListener. This commit adds a simple method that saves updatedAt via
dbal connection and refreshes the entity.
[MAILPOET-4828]

costasovo committed:
Refactor fetching subscribers stats data

In the original approach we completely ignored subscribers who subscribed and unsubscribed
within the last 30 days.
Based on the feedback from QA this was quite confusing.
This commit changes the logic and instead of looking at current status we fetch the counts
based on logs (the lastSubscribedAt column and statistics_unsubscribes table) and we ignore the current status.
On the list level we don't have any logs so we still need to check the current status on the list level,
but newly we ignore the global status.
[MAILPOET-4828]

costasovo commented:
@lysyjan I added two new commits addressing @Aschepikov notes.

lysyjan reviewed (approved):
The last changes LGTM

Aschepikov reviewed (approved):
Great, thank you, gj!
